quicktest.go

```diff
@@ -58,17 +58,45 @@ func New(t testing.TB) *C {
 type C struct {
 	testing.TB
 
-	mu       sync.Mutex
-	deferred func()
-	format   formatFunc
+	mu         sync.Mutex
+	doneNeeded bool
+	deferred   func()
+	format     formatFunc
+}
+
+// cleaner is implemented by testing.TB on Go 1.14 and later.
+type cleaner interface {
+	Cleanup(func())
 }
 
 // Defer registers a function to be called when c.Done is
 // called. Deferred functions will be called in last added, first called
-// order.
+// order. If c.Done is not called by the end of the test, the test
+// may panic. Note that if Cleanup is called, there is no
+// need to call Done.
 func (c *C) Defer(f func()) {
+	if cleaner, ok := c.TB.(cleaner); ok {
+		c.mu.Lock()
+		c.doneNeeded = true
+		c.mu.Unlock()
+		// Use TB.Cleanup when available, but add a check
+		// that Done has been called so that we don't run
+		// into unexpected Go version incompatibilities.
+		cleaner.Cleanup(func() {
+			c.mu.Lock()
+			doneNeeded := c.doneNeeded
+			c.doneNeeded = false // Prevent panicking again for the same reason.
+			c.mu.Unlock()
+			if doneNeeded {
+				panic("Done not called after Defer")
+			}
+			f()
```

Comment:
Maybe defer f() at the beginning of the function, so that the first registered cleanup function is still called on go 1.14 even if we then panic.

Reply:
Alternatively, what do you think about:
```go
if cleaner, ok := c.TB.(cleaner); ok {
		// Use TB.Cleanup when available.
		c.mu.Lock()
		doneNeeded := c.doneNeeded
		c.mu.Unlock()
		if !doneNeeded {
			// This is the first Defer call after c instantiation or a c.Done()
			// call. Add a check that Done has been called so that we don't run
			// into unexpected Go version incompatibilities. This check will be
			// run as the last cleanup function when exiting the test.
			c.mu.Lock()
			c.doneNeeded = true
			c.mu.Unlock()
			cleaner.Cleanup(func() {
				c.mu.Lock()
				doneNeeded := c.doneNeeded
				c.mu.Unlock()
				if doneNeeded {
					panic("Done not called after Defer")
				}
			})
		}
		cleaner.Cleanup(f)
		return
	}
```
Maybe easier to follow?

Reply:
done, slightly differently